### PR TITLE
View training by categories

### PR DESCRIPTION
**Pull request by chadtomkiss-madetech**
**Work done:**

- Added endpoint for getting a list of training categories with training & workers
- Updated `worker` model to associate training 
- Updated `workerTrainingCategories` model to associate mandatory training
- Updated `workerTrainingCategories` model to have a `findAllWithMandatoryTraining` method
- Updated `establishment` model to have a `findWithWorkersAndTraining` method
- Added a `trainingCategoryService` to talk to the new endpoint
- Added components to view training by category
- Added `trainingCategoryTransformer` to convert Sequelize data into a object for JSON
- Introduced `node-mocks-http` to easily mock Express requests

### CODE REVIEW

File: server/routes/workerTrainingCategories.js [hunk 1]
```diff
@@ -2,26 +2,140 @@ var express = require('express');
...
+    });
+
+    let results = trainingCategories.map((trainingCategory) => {
+      let training = []
+
+      establishment.workers.forEach((worker) => {
+        let workerTraining = worker.workerTraining.filter((workerTraining) => {
+          return workerTraining.categoryFk == trainingCategory.id
+        }).map(workerTraining => {
+          return {
```
Comment (chadtomkiss-madetech) [inline server/routes/workerTrainingCategories.js:68]: Not too keen on how this object is created

File: server/routes/workerTrainingCategories.js [hunk 2]
```diff
@@ -2,26 +2,140 @@ var express = require('express');
...
+    });
+  } catch (err) {
+    console.error(err);
+    return res.status(503).send();
+  }
+});
+
+router.route('/:establishmentId/with-training').get(async function (req, res) {
+  try {
+    let establishment = await models.establishment.findByPk(req.params.establishmentId, {
```
Comment (chadtomkiss-madetech) [inline server/routes/workerTrainingCategories.js:23]: These could be instance methods on the establishment model to reduce noise when looking at an endpoint

File: src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+  public trainingStatus(trainingRecord) {
+    return this.trainingStatusService.getTrainingStatus(trainingRecord.expires, trainingRecord.missing);
+  }
+
+  public trainingStatusCount(training, status) {
+    return training.filter((trainingRecord) => {
+        return this.trainingStatus(trainingRecord) === status;
+    }).length;
+  }
```
Comment (chadtomkiss-madetech) [inline src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.ts:66]: These could potentially live on the `TrainingStatusService`?

Comment (g-m-taylor) [reply src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.ts:66]: Are you commenting on your own pull request?!

Comment (chadtomkiss-madetech) [reply src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.ts:66]: @g-m-taylor hah! I like to comment on the things I'm not happy with just to make a note of stuff that I want to potentially amend.

It also serves as some guidance to the reviewer on the things that I think need some attention.

Comment (chadtomkiss-madetech) [reply src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.ts:66]: My friend worked on his own product and would always PR his work even though he was the only contributor - it's amazing what things you spot that you don't just looking at the code in your editor.

File: server.js [hunk 4]
```diff
@@ -180,7 +180,7 @@ app.use('/api/jobs', [refCacheMiddleware.refcache, jobs]);
 app.use('/api/localAuthority', [refCacheMiddleware.refcache, la]);
 app.use('/api/worker/leaveReasons', [refCacheMiddleware.refcache, workerLeaveReasons]);
 app.use('/api/serviceUsers', [refCacheMiddleware.refcache, serviceUsers]);
-app.use('/api/trainingCategories', [refCacheMiddleware.refcache, workingTrainingCategories]);
+app.use('/api/trainingCategories', [cacheMiddleware.nocache, workingTrainingCategories]);
```
Comment (chadtomkiss-madetech) [inline server.js:183]: @aaron-russell is it possible to only cache a specific endpoint within the `workingTrainingCategories` routes?

It was previously caching the category list which rarely changes.

However now there's an extra endpoint for training by category which I'd like not to be cached (but still cache the category list)

Comment (aaron-russell) [reply server.js:183]: Yeah, should be able to do similar to this, but use the specific endpoint.

File: server/routes/workerTrainingCategories.js [hunk 5]
```diff
@@ -1,27 +1,48 @@
...
-        order: [
-          ["seq", "ASC"]
-        ]
-      });
+      order: [['seq', 'ASC']],
+    });
 
     res.send({
-      trainingCategories: trainingCategoriesJSON(results)
+      trainingCategories: trainingCategoriesJSON(results),
```
Comment (aaron-russell) [inline server/routes/workerTrainingCategories.js:14]: You're one of those are you?

File: server/routes/workerTrainingCategories.js [hunk 6]
```diff
@@ -33,6 +54,7 @@ function trainingCategoriesJSON(givenCategories){
   });
 
   return categories;
-};
+}
```
Comment (aaron-russell) [inline server/routes/workerTrainingCategories.js:57]: So you'll add a comma but get rid of a semi colon?

File: src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.html [hunk 7]
```diff
@@ -0,0 +1,154 @@
+<div class="govuk-!-margin-bottom-6" *ngIf="showViewByToggle">
+  <a href="javascript://" (click)="this.viewTrainingByCategory.emit(false)" class="govuk-!-margin-right-6"
```
Comment (aaron-russell) [inline src/app/shared/components/training-and-qualifications-categories/training-and-qualifications-categories.component.html:2]: Using javascript in a href can cause issues with security in some browsers.

File: src/app/shared/components/training-and-qualifications-summary/training-and-qualifications-summary.component.html [hunk 8]
```diff
@@ -1,21 +1,29 @@
...
-      value="{{ sortTrainingAndQualsOption.key }}"
-    >
-      {{ sortTrainingAndQualsOption.value }}
-    </option>
-  </select>
+<div class="govuk-form-group govuk-!-margin-bottom-6">
+  <div class="govuk__flex govuk__justify-content-space-between">
+    <div *ngIf="showViewByToggle">
+      <span class="govuk-!-margin-right-6">View by staff name</span>
+      <a href="javascript://" (click)="this.viewTrainingByCategory.emit(true);" class="govuk-!-margin-right-6">View by training category</a>
```
Comment (aaron-russell) [inline src/app/shared/components/training-and-qualifications-summary/training-and-qualifications-summary.component.html:5]: Same here

File: src/app/shared/components/training-and-qualifications-summary/training-and-qualifications-summary.component.html [hunk 9]
```diff
@@ -2,7 +2,7 @@
   <div class="govuk__flex govuk__justify-content-space-between">
     <div *ngIf="showViewByToggle">
       <span class="govuk-!-margin-right-6">View by staff name</span>
-      <a href="javascript://" (click)="this.viewTrainingByCategory.emit(true);" class="govuk-!-margin-right-6">View by training category</a>
+      <a href="#" (click)="this.viewTrainingByCategory.emit(true); false" class="govuk-!-margin-right-6">View by training category</a>
```
Comment (aaron-russell) [inline src/app/shared/components/training-and-qualifications-summary/training-and-qualifications-summary.component.html:5]: Why is there just a false in the click?

**Pull request merged**